### PR TITLE
Feature/aft 13 authentication

### PR DESCRIPTION
**Pull request by gcook194**

### CODE REVIEW

File: src/main/java/com/and/digital/AndiFeedbackToolApplication.java [hunk 1]
```diff
@@ -1,8 +1,13 @@
 package com.and.digital;
 
+import com.and.digital.domain.config.SurveyMonkeyProperties;
 import org.springframework.boot.SpringApplication;
 import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.context.annotation.PropertySource;
 
+@EnableConfigurationProperties(SurveyMonkeyProperties.class)
```
Comment (AndrewBlack95) [inline src/main/java/com/and/digital/AndiFeedbackToolApplication.java]: I don't think you need @EnableConfigurationProperties, it should work if the class has @Component on it

File: src/main/java/com/and/digital/AndiFeedbackToolApplication.java [hunk 2]
```diff
@@ -1,8 +1,13 @@
...
+@PropertySource("file:/opt/applications/SurveyMonkey.properties")
```
Comment (AndrewBlack95) [inline src/main/java/com/and/digital/AndiFeedbackToolApplication.java]: Should we move this to the SurveyMonkeyProperties class? Think it's better to keep class specific annotations away from this class.

File: src/main/java/com/and/digital/domain/config/BeanConfig.java [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.web.client.RestTemplate;
+
+@Configuration
+public class BeanConfig {
+
+    @Bean
+    public RestTemplate getRestTemplate() {
+        return new RestTemplateBuilder().build();
```
Comment (AndrewBlack95) [inline src/main/java/com/and/digital/domain/config/BeanConfig.java]: We already have this in CommonConfig

File: src/main/java/com/and/digital/exception/TokenExchangeException.java [hunk 4]
```diff
@@ -0,0 +1,7 @@
+package com.and.digital.exception;
+
+public class TokenExchangeException extends RuntimeException {
+    public TokenExchangeException(String message) {
```
Comment (AndrewBlack95) [inline src/main/java/com/and/digital/exception/TokenExchangeException.java]: Missing final here

File: src/main/java/com/and/digital/service/SurveyMonkeyClient.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+import org.springframework.util.MultiValueMap;
+import org.springframework.web.client.RestTemplate;
+
+@RequiredArgsConstructor
+@Service
+public class SurveyMonkeyClient {
+
+    static final String AUTHORIZATION = "Authorization";
+    private final RestTemplate restTemplate;
+    private final SurveyMonkeyProperties properties;
```
Comment (AndrewBlack95) [inline src/main/java/com/and/digital/service/SurveyMonkeyClient.java]: We already have SurveyMonkeyRepository for the data access layer, should this code not live in there?

File: src/main/java/com/and/digital/service/SurveyMonkeyClient.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+    public String getLoginPage() {
+        final String url = loginPageUrl();
+        return restTemplate.getForObject(url, String.class);
+    }
+
+    public String exchangeShortLivedTokenForBearer(String shortLivedToken) {
```
Comment (AndrewBlack95) [inline src/main/java/com/and/digital/service/SurveyMonkeyClient.java]: Final usage missing in this method

File: src/main/java/com/and/digital/service/SurveyMonkeyClient.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+        map.add("client_secret", properties.getClientSecret());
+        map.add("code", shortLivedToken);
+        map.add("redirect_uri", properties.getSuccessfulLoginRedirectUri());
+        map.add("grant_type", properties.getGrantType());
+
+        HttpEntity<MultiValueMap<String, String>> entity = new HttpEntity<>(map, headers);
+
+        ResponseEntity<SurveyMonkeyBearerTokenResponse> response = restTemplate.exchange(properties.getOAuthTokenUrl(), HttpMethod.POST, entity, SurveyMonkeyBearerTokenResponse.class);
+
+        if (response.getStatusCode().equals(HttpStatus.OK) && response.getBody() != null) {
```
Comment (AndrewBlack95) [inline src/main/java/com/and/digital/service/SurveyMonkeyClient.java]: Extract this if to readable boolean method?

File: src/main/java/com/and/digital/service/SurveyMonkeyClient.java [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+            final String bearerToken = response.getBody().getAccessToken();
+
+            if (bearerToken != null && !bearerToken.isEmpty()) {
```
Comment (AndrewBlack95) [inline src/main/java/com/and/digital/service/SurveyMonkeyClient.java]: Can use nonNull here if you fancy

File: src/main/java/com/and/digital/web/LoginController.java [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+import javax.servlet.http.HttpServletRequest;
+
+@Slf4j
+@RequiredArgsConstructor
+@RestController
+public class LoginController {
+
+    private static final String CODE = "code";
+    private final SurveyMonkeyProperties properties;
+    private final SurveyMonkeyClient apiClient;
```
Comment (AndrewBlack95) [inline src/main/java/com/and/digital/web/LoginController.java]: No Service class? Controller should really just proxy to service classes with any logic, checks, exceptions thrown handled in the service layer

File: src/main/java/com/and/digital/web/LoginController.java [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+        if (code == null || code.isEmpty()) {
+            //TODO replace with custom exception
+            throw new IllegalArgumentException("Need a short lived code to proceed any further.");
+        }
+
+        return "";
+    }
+}
+
+
```
Comment (AndrewBlack95) [inline src/main/java/com/and/digital/web/LoginController.java:37]: Spacing here

File: src/main/java/com/and/digital/domain/config/SurveyMonkeyProperties.java [hunk 11]
```diff
@@ -2,13 +2,17 @@
 
 import lombok.Data;
 import org.springframework.boot.context.properties.ConfigurationProperties;
-import org.springframework.stereotype.Component;
 
 @Data
 @ConfigurationProperties(prefix="survey-monkey")
-@Component
 public class SurveyMonkeyProperties {
```
Comment (AndrewBlack95) [inline src/main/java/com/and/digital/domain/config/SurveyMonkeyProperties.java]: I would add @Component back as per my previous comment

File: src/main/java/com/and/digital/web/LoginController.java [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+    private final SurveyMonkeyClient apiClient;
+
+    @GetMapping("/login")
+    public ResponseEntity<String> login() {
+        final String apiResponse = apiClient.getLoginPage();
+        return ResponseEntity.ok(apiResponse);
+    }
+
+    @GetMapping("/exchange-token")
+    public String exchangeTokens(HttpServletRequest request) {
```
Comment (AndrewBlack95) [inline src/main/java/com/and/digital/web/LoginController.java]: Missing final here

File: src/main/java/com/and/digital/AndiFeedbackToolApplication.java (hunk 1 above)
Comment (gcook194) [reply src/main/java/com/and/digital/AndiFeedbackToolApplication.java]: I had made the opposite decision, so remove componant and have this, but will frevert just now.

File: src/main/java/com/and/digital/web/LoginController.java (hunk 12 above)
Comment (AndrewBlack95) [reply src/main/java/com/and/digital/web/LoginController.java]: Actually you don\'t need request here, you can just use @RequestParam to pull the param off the request:

https://www.baeldung.com/spring-request-param

File: src/main/java/com/and/digital/service/SurveyMonkeyClient.java [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+                return bearerToken;
+            }
+        }
+
+        throw new TokenExchangeException("Could not exchange short lived token for bearer token");
+    }
+
+    private String loginPageUrl() {
```
Comment (AndrewBlack95) [inline src/main/java/com/and/digital/service/SurveyMonkeyClient.java]: Wondering if it's better to build this in the properties class ? Means you don't need to build this every time we get a request and it seems like properties code to me

File: src/main/java/com/and/digital/service/SurveyMonkeyClient.java [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+import org.springframework.stereotype.Service;
+import org.springframework.util.LinkedMultiValueMap;
+import org.springframework.util.MultiValueMap;
+import org.springframework.web.client.RestTemplate;
+
+@RequiredArgsConstructor
+@Service
+public class SurveyMonkeyClient {
+
+    static final String AUTHORIZATION = "Authorization";
```
Comment (AndrewBlack95) [inline src/main/java/com/and/digital/service/SurveyMonkeyClient.java]: Spring Boot have a constant for this: HttpHeaders.AUTHORIZATION

File: src/main/java/com/and/digital/service/SurveyMonkeyClient.java [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+        final String url = loginPageUrl();
+        return restTemplate.getForObject(url, String.class);
+    }
+
+    public String exchangeShortLivedTokenForBearer(String shortLivedToken) {
+
+        HttpHeaders headers = new HttpHeaders();
+        headers.setContentType(MediaType.APPLICATION_FORM_URLENCODED);
+
+        MultiValueMap<String, String> map = new LinkedMultiValueMap<>();
```
Comment (AndrewBlack95) [inline src/main/java/com/and/digital/service/SurveyMonkeyClient.java]: Out of curiosity, why did you go for a MultiValueMap?

File: src/test/java/com/and/digital/service/SurveyMonkeyClientTest.java [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+
+        properties = new SurveyMonkeyProperties();
+        properties.setClientId("some-client-id");
+        properties.setClientSecret("some-client-secret");
+        properties.setOAuthTokenUrl("https://some-url.com/token");
+        properties.setSuccessfulLoginRedirectUri("https://some-rul.com/home");
+        properties.setGrantType("authorization_code");
+
+        clientUnderTest = new SurveyMonkeyClient(restTemplate, properties);
+    }
```
Comment (AndrewBlack95) [inline src/test/java/com/and/digital/service/SurveyMonkeyClientTest.java]: Think you are missing tests for like response body being null. status code not being OK and access token being null

File: src/test/java/com/and/digital/web/LoginControllerTest.java [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+    private static final String CLIENT_SECRET = "secret-1234";
+
+    @BeforeEach
+    void setup() {
+        surveyMonkeyProperties = new SurveyMonkeyProperties();
+        surveyMonkeyProperties.setClientId(CLIENT_ID);
+        surveyMonkeyProperties.setClientSecret(CLIENT_SECRET);
+
+        controllerUnderTest = new LoginController(surveyMonkeyProperties, surveyMonkeyClient);
+    }
```
Comment (AndrewBlack95) [inline src/test/java/com/and/digital/web/LoginControllerTest.java:26]: Seems like you are missing a negative test for login()

File: src/test/java/com/and/digital/web/LoginControllerTest.java [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+
+    @Test
+    void login_ReturnsLoginUrl() {
+
+        ResponseEntity<String> result = controllerUnderTest.login();
+
+        assertThat(result.getStatusCode(), is(HttpStatus.OK));
```
Comment (AndrewBlack95) [inline src/test/java/com/and/digital/web/LoginControllerTest.java:38]: This test is failing on my machine

File: src/main/java/com/and/digital/service/SurveyMonkeyClient.java (hunk 8 above)
Comment (gcook194) [reply src/main/java/com/and/digital/service/SurveyMonkeyClient.java]: You'll need to explain this one mate, never used it before.

File: src/main/java/com/and/digital/service/SurveyMonkeyClient.java (hunk 13 above)
Comment (gcook194) [reply src/main/java/com/and/digital/service/SurveyMonkeyClient.java]: we're likely going to remove it

File: src/test/java/com/and/digital/web/LoginControllerTest.java (hunk 18 above)
Comment (gcook194) [reply src/test/java/com/and/digital/web/LoginControllerTest.java:38]: sorted now

File: src/main/java/com/and/digital/service/SurveyMonkeyClient.java (hunk 8 above)
Comment (AndrewBlack95) [reply src/main/java/com/and/digital/service/SurveyMonkeyClient.java]: It's just a more readable version of a null check

File: src/main/java/com/and/digital/service/SurveyMonkeyClient.java (hunk 14 above)
Comment (gcook194) [reply src/main/java/com/and/digital/service/SurveyMonkeyClient.java]: Removing it for now.

File: src/main/java/com/and/digital/service/SurveyMonkeyClient.java (hunk 15 above)
Comment (gcook194) [reply src/main/java/com/and/digital/service/SurveyMonkeyClient.java]: Just looked on Google for how to send a form encoded post from Java and that's what it came back with.
